### ISSUE TITLE
rustc stack overflow: recursive associated types

### ISSUE THREAD

**Issue by seanmonstar** [rust-lang/rust#23992]
```rust
trait Foo {
    type Bar;
    fn foo(self) -> Self::Bar;
}

impl Foo for Box<Foo> {
    type Bar = <Self as Foo>::Bar;
    fn foo(self) -> <Self as Foo>::Bar {
        (*self).foo()
    }
}
```

**Comment by apoelstra** [comment]
I've encountered a variant of this bug (or is it another bug? please advise if so and I will post it separately):
````
pub struct Receipt<T: Trait>(T);
pub struct Future<'a>(&'a bool);

pub trait Trait {
    type Error;
    fn ready(self) -> Self::Error;
}

impl<'a> Trait for Future<'a> {
    type Error = Receipt<Future<'a>>;
    fn ready(mut self) -> Receipt<Future<'a>> {}
}
````
If I remove the lifetime, the stack overflow no longer occurs.

**Comment by nham** [comment]
Seems to be a dupe of #21946

**Comment by arielb1** [comment]
@apoelstra's example was fixed in 1.4.
